### ISSUE TITLE
Line breaks are not removed

### ISSUE THREAD

**Issue by Techn1c4l** [kristian/minify-xml#15]
While using this package I found out that it doesn't remove line breaks. Since XML does not require a specific form of line break for correct documents, this would be very handy. Most other minification libraries for JSON, CSS, etc. in fact remove line breaks.
Right now I use this workaround. This does keep text in CDATA sections intact.

`MinifyXML.minify(input, { considerPreserveWhitespace: false, shortenNamespaces: false }).replace(/(?<cdata>\<!\[CDATA\[.*?]]>)|(?:\r?\n)/gs, (match: string, cdata: any) => cdata ?? '');`

**Comment by kristian** [comment]
Hey @Techn1c4l, can you provide an example XML which line breaks you actually mean? `minify-xml` attempts remove all line breaks that are not syntactically changing the document. This is determined by removing / trimming non-empty text nodes. E.g.:

```
<tag>

<tag2>
```

Is minifier to `<tag><tag2>`. The only line breaks / whitespaces that are not trimmed, are those inside of a non-empty text node, like:

```
<tag>Test

Hello</tag>
```

Because removing the whitespaces would actually change the document content to: `<tag>TestHello</tag>`, which is not an acceptable minification result. Can you elaborate what whitespaces you specifically want to have removed?

**Comment by Techn1c4l** [comment]
After tinkering a bit, I've managed to retrieve the example which I can reproduce this on.
```
<?xml version="1.0" encoding="UTF-8">

<root>


  <test>
123

  </test>


  <r><![CDATA[

   characters with markup

]]></r></root>
```
This minifies to:
```
<?xml version="1.0" encoding="UTF-8">

<root><test>
123

  </test><r><![CDATA[

   characters with markup

]]></r></root>
```
I'm aware that this is not syntaxicaly correct, but I still believe line breaks before opening < root > should be removed.

**Comment by kristian** [comment]
The whitespace before <root> shoult totally be removed, because it doesn‘t semantically change the document. I would even consider this a bug at this point in time. I think a test case is missing for this. Thanks for bringing it up, I will look into this asap.

**Comment by kristian** [comment]
What I don‘t understand is that in [this test case](https://github.com/kristian/minify-xml/blob/master/test/prolog_doctype/in.xml) whitespaces are properly removed. Could you create a PR that shows a failing test? Best to create a `issue_15` folder in the test directory providing an input file and the expected outcome. Thanks.

**Comment by Techn1c4l** [comment]
I'll try, just not so familiar with all this git/PR stuff. The main difference is my example doesn't have proper ?> closing in the XML declaration.

While I'm trying to figure out how to do the PR, check https://www.atatus.com/tools/xml-minify . I believe it's using your library.

Here's the example:
```
<?xml version="1.0" encoding="UTF-8">

<root><test>
123

  </test><r><![CDATA[

   characters with markup

]]></r></root>
```

Now this minifies correctly:
```
<?xml version="1.0" encoding="UTF-8"?>

<root><test>
123

  </test><r><![CDATA[

   characters with markup

]]></r></root>
```

P.S. I tried your example and if I replace ?> with > in the declaration, it fails to remove the whitespace.

**Comment by kristian** [comment]
Thanks for the effort. I will take a look as soon as I am back home / latest tomorrow. Shouldn‘t be a big issue though! Thanks again.

btw. [`minify-x.ml`](https://minify-x.ml/) is not only a library, but also a website :)

**Comment by kristian** [comment]
Just to confirm. Yes indeed my minifier will not remove the whitespace in the prolog if the prolog tag isn‘t closed with [propper syntax](https://www.w3.org/TR/xml/#sec-prolog-dtd). `>` (without a `?`) is not a valid syntax and my minifier will ignore it. However I agree that making the minifier behave more lenient in those cases, because it doesn‘t change the minification attempt. However I will remove the bug label and classify this as an improvement instead. Expect a fix tomorrow please, I am not at home right now. Thanks.

**Comment by Techn1c4l** [comment]
Totally agree. This is indeed not a bug, but it would be nice if your library would do the best effort to minify given XML string. I believe only CDATA sections should be left intact, whitespaces in any other places should get removed.

**Comment by kristian** [comment]
In any other *valid* place, I agree. Whitespace in text blocks will only be removed if you set the `trimWhitespaceFromTexts` and / or `collapseWhitespaceInTexts` option, which will potentially change the meaning of the XML document (which is why the options are `false` by default). Any other optimization that does not alter the content / meaning of the document, I agree. However my library still has to „respect“ the general XML syntax, as different types of sections in XML (like prolog, CDATA, etc.) need different rules for minifcation.

**Comment by kristian** [comment]
Fixed with [v3.15.0](https://github.com/kristian/minify-xml/releases/tag/3.5.0)
